Add gptel-fn-complete

mwolson opened the pull request:
### Brief summary of what the package does

Complete the function at point using gptel.

This uses the existing `gptel-rewrite.el` library to perform completion on an
entire function, replacing what's already written so far in that function in
a way that prefers to complete the end of the function, but may also apply
small changes to the original function.

### Direct link to the package repository

https://github.com/mwolson/gptel-fn-complete

### Your association with the package

I'm the maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] (used melpazoid) I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass -

I do see in your README the following:

> This is intended to be more of a tech demo than a final project

I suspect this statement is probably out of date - but if you still view this as a tech demo, I sometimes recommend to authors that they let the package mature over a month or two more before being publicly indexed (breaking changes can be more painful when there's an established userbase).

Otherwise this looks good; and in case you haven't, ensure GitHub release v0.2.0 is up-to-date with your current code and `Package-Version`.

riscy commented:
Actually, while I'm here, I wonder if you have had a chance to compare to the following packages:

- https://melpa.org/#/elysium
- https://melpa.org/#/minuet

mwolson commented:
Hi,

I've removed the tech demo verbiage and added some comparisons to other tools in the README.md, as well as a new v0.3.0 release for some unrelated changes.

The comparison - excerpted from README.md - is currently:

* [aidermacs](https://github.com/MatthewZMD/aidermacs) - `aidermacs` works with
  a commandline helper tool named `aider` which does most of the heavy
  lifting. It can record system prompts and conversations to files in your
  project to better remember what needs to be done. It has significantly more
  capabilities.

  `gptel-fn-complete` is built on `gptel` and is very minimally targeted towards
  just completing a single function, paragraph, or region.

* [elysium](https://github.com/lanceberge/elysium) - `elysium` sends either the
  entire file or a selected region to the LLM. It also shows a window on the
  side with the LLM response, and includes smerge markers into the buffer with
  the result from the LLM. It is built on `gptel`.

  `gptel-fn-complete` uses a standard `gptel-rewrite` overlay on the code result
  instead of using the more visually-busy smerge markers. It can be configured
  using `(setopt gptel-rewrite-default-action 'accept)` to automatically skip the
  `gptel-rewrite` overlay and directly insert the text. It does not change the
  window layout to show a chat buffer.

* [minuet](https://github.com/milanglacier/minuet-ai.el) - `minuet` typically
  does a "fill-in-the-middle" style completion that doesn't modify the code
  before the point. It also has an optional auto-suggestion capability as you
  type. It can automatically include additional context lines before or after
  the function. It is not built on `gptel` and must be configured separately.

  `gptel-fn-complete` allows a partial rewrite of the entire function, if the
  LLM thinks there is stylistic value in the rewrite, and is typically manually
  triggered by a keybind. It does not have an auto-suggestion mode. It also
  leverages your existing `gptel` setup, allowing tricks like using gptel to
  manually include more file context to advise a better completion.

riscy commented:
Thanks for that informative reply.  In case you haven't, ensure GitHub release v0.3.1 is up-to-date with your current code and `Package-Version` - but I will merge now.

mwolson commented:
Thanks! Yep the v0.3.1 version tag corresponds with a commit to bump the `Version` header to `0.3.1` which is hopefully also enough to get it into MELPA Stable.
